Proposal: Add First and Single extension methods for collection checks

ocoanet opened the pull request:
I regularly need to check properties on the first or single element of a collection.
It is easy to write:

``` cs
Check.That(enumerable.First()).IsEqualTo(42);
```

Yet I would like the exception to be generated from an assertion and not from the `First` method.
My tests can use custom `ExpectedFirst` and `ExpectedSingle` methods but those do not really follow the NFluent philosophy.

So I propose to add extension methods for collection checks that could be used in the following way:

``` cs
Check.That(enumerable).First().IsEqualTo(42);
```

dupdob commented:
Thanks for your input. This is a pull request of great quality.
I understand why Check.That(...).First() is better than Check.That(...First"()), but it looks like there is a need to support subsequent items as well, don't you think?

ocoanet commented:
Do you mean that other collection checks should be included, such as `ElementAt` or `Last`?

rhwy commented:
If it is that, I think it could be added later, when someone feel the need for it. once the base, when you have the code for First() is here it's easy to add later the code for Last(). Because if we start looking in that direction, it should also be reasonable to implement all the enumerable extensions. ;-)

tpierrain commented:
Excellent pull request Olivier, thanks for submitting it. We'll definitely add such feature (First, Last, HasOneElementOnly,...) but I'd like few extra days to think about the naming/semantic/grammar part if you don't mind. I'd like to be sure that we will introduce something that still fit with the natural language. We'll get back to you ASAP on this!

tpierrain commented:
For instance, what do you think about:

    Check.That(enumerable).HasFirstElement().Which.IsEqualTo(42);
    Check.That(enumerable).HasLastElement().Which.IsGreaterThan(10);

But also

    Check.That(enumerable).HasElementNumber(3).Which.IsNull();

And 

    Check.That(enumerable).HasOneElementOnly();

tpierrain commented:
Also, we could also combine the last suggestion with a __That__ element. Something like:

    Check.That(enumerables).HasOneElementOnly().That.IsGreaterThan(3);

ocoanet commented:
I like the syntax of the last suggestion. I suppose that it would imply to add a new interface like `ICollectionElementCheck` or `ICheckProviderCheck` to expose the `That` property. I can try to do that.

tpierrain commented:
Absolutely. 

If you want -and if there is no rush- I can implement it, but I'd be very happy to have your contribution on this.

As a source of inspiration, you can see the INullableOrNumberCheckLink interface (where the "Which" property/operator is currently handled)

Just let me know if you want me to implement it or not.

And again, thanks for using NFluent and making it better via your feedbacks and contributions. 

Cheers

ocoanet commented:
Sorry for my very late reply. I wish I would have had the time to do it. I might be able to work on the feature this week. Have you started something yet?

tpierrain commented:
Not yet, since I understood that you would like to do it. But I can have a look at it if you want.

ocoanet commented:
I finally took the time to update my PR ! I implemented the suggested extension methods and added a new type, `ICollectionElementCheck<T>`, to improve the check syntax.

The code is really simple and I see a few possible improvements:
- `ICheckLink` could be implemented.
- The extension methods could be moved to an other class because `EnumerableCheckExtensions` is getting quite big.
- The method names could be changed to match Linq methods: `HasSingleElement` instead of `HasOneElementOnly` and `HasElementAt` instead of `HasElementNumber`.
- I wrote custom code for `IList<T>` that could be considered as premature optimization and removed.

Tell me if you want some changes, I will try to be more responsive this time :)

dupdob commented:
Hi @ocoanet, I am currently reviewing your PR for integration (at last).
I am dealing with the merge conflicts, due to the lead time (sorry about that).
Here is what I will change:
- index will start at 0 (instead of 1)
- linking word is 'Which' (instead of your proposal for 'That'), for consistency reason 
- I'll add support for negation (Not)
